README.md
redis-hashes
===============

[![Build Status](https://travis-ci.org/lykmapipo/redis-hashes.svg?branch=master)](https://travis-ci.org/lykmapipo/redis-hashes)
[![Dependency Status](https://img.shields.io/david/lykmapipo/redis-hashes.svg?style=flat)](https://david-dm.org/lykmapipo/redis-hashes)
[![npm version](https://badge.fury.io/js/redis-hashes.svg)](https://badge.fury.io/js/redis-hashes)

redis hash utilities for nodejs

## Requirements
- [Redis 2.8.0+](http://redis.io/)
- [NodeJS 8.1.4+](https://nodejs.org/en/)

## Installation
```sh
$ npm install --save redis-hashes
```

## Usage

```javascript
//initialize redis-hashes with default options
const redis = require('redis-hashes')([options]);

//save single user
const user = ...;
hash.save(user, {collection: 'users' }, done);

//save multiple users
const users = ...;
hash.save(users, {collection: 'users' }, done);


//get single user
hash.get(<id>, function(error, user){
   ...
});

//get single user and select specified fields
hash.get(<id>, { fields: 'name, email'}, function(error, user){
   ...
});


//get multiple users
hash.get(<id>, <id>, function(error, users){
   ...
});

//search users collection
hash.search(<search_query>, function (error, users) {
    ...
});


//remove single users
hash.remove(<id>, function(error, results){
   ...
});


//remove multiple users
hash.remove(<id>, <id>, function(error, results){
   ...
});

...

```

## Options
- `prefix:String` - redis key prefix. default to `r`
- `separator:String` - redis key separator. default to `:`
- `redis:Object` - [redis](https://github.com/NodeRedis/node_redis#rediscreateclient) connections options.

To initialize `redis` with custom options use

```js
const redis = require('redis-hashes')({
    prefix:'q',
    separator:'-',
    redis: {
    port: 6379,
    host: '127.0.0.1'
  }
});

...

```

## API

### `save(objects:Object|Object[],[options:Object],done:Fuction)`
Store`(create or update)` object(s) into redis [hash datatype](http://redis.io/commands/HSET). Before saving the whole of object is [flatten'ed](https://github.com/hughsk/flat) and serialized. i.e all `dates` will be converted to `timestamps` etc.


Options:
- `index:Boolean` - whether to [index](https://github.com/tj/reds) the object or not for search. default to `true`.
- `collection:String` - name of collection or namespace used in prefix hash keys. default to `hash`
- `ignore: Array[String]` - Collection of `object fields` to ignore when indexing 

```js
const object = ...;
hash.save(object, function (error, saved) {
    ...
});

const user = ...;
hash.save(user, {collection: 'users' }, function (error, saved) {
    ...
});

//or bulk save
const objects = ...;
hash.save(objects, function (error, objects) {
    ...
});

const users = ...;
hash.save(users, {collection: 'users' }, function (error, users) {
    ...
});
```

### `get(...keys,[{ fields: String|String[] }],done:Function)`
Get single or multiple saved object using their keys

```js
//get single
hash.get(<id>, function(error, object){
   ...
});

//get single
hash.get(<id>, { fields: 'name, email'}, function(error, object){
   ...
});

//get multiple object
hash.get([<id>, <id>], function(error, objects){
   ...
});

//get multiple object
hash.get([<id>, <id>], { fields: 'name, email'},  function(error, objects){
   ...
});

//get multiple object
hash.get(<id>, <id>, function(error, objects){
   ...
});

//get multiple object
hash.get(<id>, <id>, { fields: 'name, email'}, function(error, objects){
   ...
});
```

### `search(options:String|Object,[{ fields: String|String[]}],done:Function)`
Search existing objects.

Options:
- `type:String` - type of [reds](https://github.com/tj/reds) search. default to `or`
- `collection:String` - name of collection used in searching. default to `hash`
- `q: String` - query string. default to ''

```js
//search default collection
hash.search(<search_query>, function (error, objects) {
    ...
});

//search default collection and select specified fields
hash
  .search({q:<search_query>, fields: 'name, email'}, function (error, objects) {
    ...
});

//search specific collection
hash.search({
    q: <search_query>,
    collection: 'users',
    type:'or'
  }, function (error, objects) {
    ...
});

//search specific collection and select specified fields
hash.search({
    q: <search_query>,
    collection: 'users',
    type:'or',
    fields: 'name, email'
  }, function (error, objects) {
    ...
});
```

### `remove(...keys,done:Function)`
Remove single or multiple saved object using their keys

```js
//remove single
hash.remove(<id>, function(error, object){
   ...
});

//remove multiple object
hash.remove([<id>, <id>], function(error, objects){
   ...
});


//remove multiple object
hash.remove(<id>, <id>, function(error, objects){
   ...
});
```

## References
- [store-javascript-objects-in-redis-with-node-js-the-right-way](https://medium.com/@stockholmux/store-javascript-objects-in-redis-with-node-js-the-right-way-1e2e89dbbf64#.eb1040111)
- [reds-internals-searching-and-better-searching-with-node-js-and-redis](https://medium.com/@stockholmux/reds-internals-searching-and-better-searching-with-node-js-and-redis-57da99077e83#.5nhdaxnl4)


## Testing
* Clone this repository

* Install all development dependencies
```sh
$ npm install
```

* Then run test
```sh
$ npm test
```

## Contribute
It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## Licence

The MIT License (MIT)

Copyright (c) 2017 lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 

CHANGELOG.md
# CHANGELOGS

## 0.1.0 - 2016-09-03
- Initial implementation

## 0.2.0 - 2016-09-04
- Allow selecting specific fields from hash.get and hash.search

## 0.3.0 - 2016-09-04
- Allow removing specified hash by key

## 0.4.0 - 2016-09-04
- Add timestamps(createdAt & updatedAt) to the saved object
- Ensure update & save update timestamps
- Allow multisave
- Improve code JSDOCS

test/bootstrap.spec.js
'use strict';

process.env.NODE_ENV = 'test';

//global dependencies
const redis = require('redis-clients')();

before(function (done) {
  redis.clear(done);
});


before(function () {
  redis.reset();
});


after(function (done) {
  redis.clear(done);
});


after(function () {
  redis.reset();
});

test/hash.unit.spec.js
'use strict';

//dependencies
const path = require('path');
const _ = require('lodash');
const expect = require('chai').expect;
const faker = require('faker');
const redis = require('redis-clients')();
const hash = require(path.join(__dirname, '..'))();


describe('hash', function () {

  before(function () {
    redis.reset();
  });

  before(function () {
    redis.clear();
  });

  before(function () {
    redis.init();
  });

  describe('serialize & deserialize', function () {
    it('should be able to serialize dates to timestamp', function () {
      const today = new Date();
      const object = { today: today };
      const serialized = hash.serialize(object);
      expect(serialized.today).to.be.equal(today.getTime());
    });
  });


  describe('save', function () {
    before(function () {
      redis.clear();
    });

    let _id;
    let saved;
    const object = hash.deserialize(faker.helpers.userCard());
    const objects = [
      hash.deserialize(faker.helpers.userCard()),
      hash.deserialize(faker.helpers.userCard()),
    ];

    it('should be able to able to save object to redis', function (done) {
      hash.save(object, function (error, _object) {
        _id = _object._id;
        saved = _object;
        expect(_object.createdAt).to.exist;
        expect(_object.updatedAt).to.exist;
        _object =
          _.omit(_object, '_id', 'createdAt', 'updatedAt');
        expect(_object).to.be.eql(object);
        done(error, _object);
      });
    });

    it(
      'should be able to save object into using different collection',
      function (done) {
        hash.save(object, { collection: 'users' }, function (
          error, _object) {
          _id = _object._id;
          expect(_id).to.contain('users');
          expect(_object.createdAt).to.exist;
          expect(_object.updatedAt).to.exist;
          _object =
            _.omit(_object, '_id', 'createdAt', 'updatedAt');
          expect(_object).to.be.eql(object);
          done(error, _object);
        });
      });

    it('should be able to able to update existing object in redis',
      function (done) {
        hash.save(saved, function (error, _object) {
          expect(_object._id).to.be.equal(saved._id);
          expect(_object.createdAt).to.exist;
          expect(_object.updatedAt).to.exist;
          expect(_object.createdAt).to.be.equal(saved.createdAt);
          expect(_object.updatedAt).to.not.be.equal(saved.updatedAt);
          expect(_object.updatedAt > saved.updatedAt).to.be.true;
          _object =
            _.omit(_object, '_id', 'createdAt', 'updatedAt');
          expect(_object).to.be.eql(object);
          done(error, _object);
        });
      });

    it('should be able to able to save multiple objects to redis',
      function (done) {
        hash.save(objects, function (error, _objects) {
          let _object = _.first(_objects);
          expect(_object.createdAt).to.exist;
          expect(_object.updatedAt).to.exist;
          _object =
            _.omit(_object, '_id', 'createdAt', 'updatedAt');
          expect(_object).to.be.eql(_.first(objects));
          done(error, _object);
        });
      });

  });

  describe('get', function () {
    before(function () {
      redis.clear();
    });

    let _idx, _idy;
    let objectx = hash.deserialize(faker.helpers.userCard());
    let objecty = hash.deserialize(faker.helpers.userCard());
    const fields = ['name', 'email', 'phone', 'company.name'];

    before(function (done) {
      hash.save(objectx, function (error, _object) {
        _idx = _object._id;
        done(error, _object);
      });
    });

    before(function (done) {
      hash.save(objecty, function (error, _object) {
        _idy = _object._id;
        done(error, _object);
      });
    });

    it('should be able to fetch single object', function (done) {
      hash.get(_idx, function (error, _object) {
        expect(_object._id).to.be.equal(_idx);
        expect(_object.createdAt).to.exist;
        expect(_object.updatedAt).to.exist;
        _object =
          _.omit(_object, '_id', 'createdAt', 'updatedAt');
        expect(_object).to.be.eql(objectx);
        done(error, _object);
      });
    });

    it(
      'should be able to fetch only specified fields of a single object',
      function (done) {
        hash.get(_idx, { fields: fields }, function (error, _object) {
          expect(_object._id).to.be.equal(_idx);
          expect(_object.name).to.be.equal(objectx.name);
          expect(_object.email).to.be.equal(objectx.email);
          expect(_object.phone).to.be.equal(objectx.phone);
          expect(_object.company.name)
            .to.be.equal(objectx.company.name);
          done(error, _object);
        });
      });

    it(
      'should be able to fetch only specified fields of a single object',
      function (done) {
        hash.get(_idx, { fields: 'name' }, function (error, _object) {
          expect(_object._id).to.be.equal(_idx);
          expect(_object.name).to.be.equal(objectx.name);
          done(error, _object);
        });
      });

    it(
      'should be able to fetch only specified fields of a single object',
      function (done) {
        hash.get(_idx, { fields: 'name, email' }, function (error,
          _object) {
          expect(_object._id).to.be.equal(_idx);
          expect(_object.name).to.be.equal(objectx.name);
          expect(_object.email).to.be.equal(objectx.email);
          done(error, _object);
        });
      });

    it('should be able to fetch multiple objects', function (done) {
      hash.get([_idx, _idy], function (error, objects) {
        expect(error).to.not.exist;
        expect(objects).to.have.have.length(2);
        done(error, objects);
      });
    });

    it('should be able to fetch multiple objects', function (done) {
      hash.get(_idx, _idy, function (error, objects) {
        expect(error).to.not.exist;
        expect(objects).to.have.have.length(2);
        done(error, objects);
      });
    });

    it(
      'should be able to fetch multiple object with only specified fields',
      function (done) {
        hash.get(_idx, _idy, { fields: fields }, function (error,
          objects) {
          expect(error).to.not.exist;
          expect(objects).to.have.have.length(2);
          done(error, objects);
        });
      });

  });

  describe('search', function () {
    before(function () {
      redis.clear();
    });

    let _idx, _idy;
    let objectx = faker.helpers.userCard();
    let objecty = faker.helpers.userCard();

    before(function (done) {
      hash.save(objectx, function (error,
        _object) {
        _idx = _object._id;
        expect(_object.createdAt).to.exist;
        expect(_object.updatedAt).to.exist;
        _object =
          _.omit(_object, '_id', 'createdAt', 'updatedAt');
        objectx = _object;
        done(error, _object);
      });
    });

    before(function (done) {
      hash.save(objectx, { collection: 'users' }, function (
        error, _object) {
        _idy = _object._id;
        expect(_object.createdAt).to.exist;
        expect(_object.updatedAt).to.exist;
        _object =
          _.omit(_object, '_id', 'createdAt', 'updatedAt');
        objecty = _object;
        done(error, _object);
      });
    });

    it('should be able to search saved objects', function (done) {
      hash.search(objectx.username, function (error, objects) {
        expect(error).to.not.exist;
        expect(objects).to.have.have.length(1);

        let _object = objects[0];
        expect(_object.createdAt).to.exist;
        expect(_object.updatedAt).to.exist;
        _object =
          _.omit(_object, '_id', 'createdAt', 'updatedAt');
        expect(_object).to.be.eql(objectx);
        done(error, objects);
      });
    });

    it('should be able to search a specific collection', function (done) {
      hash.search({
        q: objectx.username,
        collection: 'users'
      }, function (error, objects) {
        expect(error).to.not.exist;
        expect(objects).to.have.have.length(1);

        let _object = objects[0];
        expect(_object.createdAt).to.exist;
        expect(_object.updatedAt).to.exist;
        _object =
          _.omit(_object, '_id', 'createdAt', 'updatedAt');
        expect(_object).to.be.eql(objecty);
        done(error, objects);
      });
    });

    it(
      'should be able to search saved objects and select specified fields',
      function (done) {
        hash
          .search({ q: objectx.username, fields: 'name, email' },
            function (error, objects) {
              expect(error).to.not.exist;
              expect(objects).to.have.have.length(1);
              let _object = objects[0];
              _object =
                _.omit(_object, '_id', 'createdAt', 'updatedAt');
              expect(_object.name).to.be.eql(objectx.name);
              done(error, objects);
            });
      });
  });

  describe('remove', function () {
    let _id;
    const object = hash.deserialize(faker.helpers.userCard());

    before(function () {
      redis.clear();
    });

    before(function (done) {
      hash.save(object, function (error, _object) {
        _id = _object._id;
        done(error, _object);
      });
    });

    it(
      'should be able to delete existing object',
      function (done) {
        hash.remove(_id, function (error, result /*should it be deleted object*/ ) {
          if (error) { done(error); }

          //continue with assertions
          expect(result).to.exists;
          //ensure doesnt exists
          hash.get(_id, function (error, found) {
            expect(error).to.exist;
            expect(error.status).to.be.equal(404);
            expect(found).to.not.exist;
            done();
          });
        });
      });

  });

  after(function (done) {
    redis.clear(done);
  });

  after(function () {
    redis.reset();
  });

});

test/query.unit.spec.js
'use strict';


//global dependencies
const path = require('path');
const expect = require('chai').expect;
const query = require(path.join(__dirname, '..', 'src', 'query'));


describe('query', function () {

  describe('select', function () {

    it('should be able to parse fields to select', function () {
      expect(query.select(['a', 'b'])).to.be.eql(['_id', 'a', 'b']);
      expect(query.select('a,b')).to.be.eql(['_id', 'a', 'b']);
      expect(query.select('a, b')).to.be.eql(['_id', 'a', 'b']);
      expect(query.select('a, b ')).to.be.eql(['_id', 'a', 'b']);
      expect(query.select(' a, b ')).to.be.eql(['_id', 'a', 'b']);
      expect(query.select(' a , b ')).to.be.eql(['_id', 'a', 'b']);

      expect(query.fields(['a', 'b'])).to.be.eql(['_id', 'a', 'b']);
      expect(query.fields('a,b')).to.be.eql(['_id', 'a', 'b']);
      expect(query.fields('a, b')).to.be.eql(['_id', 'a', 'b']);
      expect(query.fields('a, b ')).to.be.eql(['_id', 'a', 'b']);
      expect(query.fields(' a, b ')).to.be.eql(['_id', 'a', 'b']);
      expect(query.fields(' a , b ')).to.be.eql(['_id', 'a', 'b']);
    });

  });

});

Gruntfile.js
'use strict';

module.exports = function (grunt) {

  // add grunt tasks.
  grunt.loadNpmTasks('grunt-mocha-test');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');

  grunt.initConfig({
    // Configure a mochaTest task
    mochaTest: {
      unit: {
        options: {
          reporter: 'spec',
          timeout: 20000
        },
        src: [
          'test/**/*.js'
        ]
      },
      intergration: {
        options: {
          reporter: 'spec',
          timeout: 20000
        },
        src: [
          'test/**/*.js',
        ]
      }
    },
    jshint: {
      options: {
        reporter: require('jshint-stylish'),
        jshintrc: '.jshintrc'
      },
      all: [
        'Gruntfile.js',
        'index.js',
        'src/**/*.js',
        'test/**/*.js'
      ]
    },
    watch: {
      all: {
        files: [
          'Gruntfile.js',
          'index.js',
          'src/**/*.js',
          'test/**/*.js'
        ],
        tasks: ['default']
      }
    }
  });

  //custom tasks
  grunt.registerTask('default', ['jshint', 'mochaTest:unit', 'watch']);
  grunt.registerTask('test', ['jshint', 'mochaTest:unit']);

};

src/query.js
'use strict';

/**
 * @module query
 * @name query
 * @description hash query utilities
 * @author lally elias<lallyelias87@gmail.com>
 * @singleton
 * @public
 * @since 0.1.0
 * @version 0.1.0
 */


//global dependencies
const _ = require('lodash');


//NOTE!: all hash has implicit _id field as an id field


/**
 * @function
 * @name select
 * @param  {Array|String} fields field to select from redis hash
 * @example
 * 	['name', 'age', 'company.name'] or
 * 	'name, age, company.name' or
 * 	'name,age,company.name'
 * @return {Array} unique fields to be selected from hash
 * @since  0.2.0
 * @version 0.1.0
 * @public
 */
exports.select = exports.fields = function (fields) {

  //ensure fields
  fields = _.compact(['_id'].concat(fields));

  //clean fields
  fields = _.chain(fields)
    .map(function (field) {
      //split comma separated strings
      return _.split(field, ',');
    })
    .flattenDeep() //flat all fields into array
    .map(function (field) {
      //trim fields
      return _.trim(field);
    })
    .compact().uniq().value(); //ensure unique fields

  return fields;

};

src/hash.js
'use strict';


/**
 * @module redis-hash
 * @name redis-hash
 * @description redis hash utilities
 * @author lally elias<lallyelias87@gmail.com>
 * @singleton
 * @public
 * @version 0.4.1
 * @see  {@link https://github.com/lykmapipo/redis-hash}
 */


//global dependencies
const path = require('path');
const _ = require('lodash');
const async = require('async');
const traverse = require('traverse');
const uuid = require('uuid');
const reds = require('reds');
const flat = require('flat').flatten;
const unflat = require('flat').unflatten;
const redis = require('redis-clients')();


//local dependencies
const query = require(path.join(__dirname, 'query'));


//initialize defaults settings
const defaults = {
  prefix: 'r',
  separator: ':',
  redis: {
    port: 6379,
    host: '127.0.0.1'
  }
};


//default fields to be ignored from search indexes
const ignore = ['_id', 'createdAt', 'updatedAt'];


/**
 * @name defaults
 * @description default redis client connection options
 * @type {Object}
 * @since 0.1.0
 * @public
 */
exports.defaults = _.merge({}, defaults, redis.defaults);


/**
 * @function
 * @name serialize
 * @description traverse js object and try convert values to appropriate redis
 *              redis storage type
 *              e.g dates will be converted to timestamp
 * @param  {Object} object valid js plain object
 * @return {Object}        object with all nodes converted to their respective
 *                         possible redis storage types
 *
 * @since 0.1.0
 * @private
 */
exports.serialize = function (object) {

  //ensure object
  object = _.merge({}, object);

  //TODO allow custom serializers

  //traverse object and apply serializers
  traverse(object).forEach(function (value) {

    //parse convert dates to timestamps
    const isDate = _.isDate(value);
    if (isDate) {
      value = value.getTime();
      //update current field
      this.update(value);
    }

  });

  return object;
};


/**
 * @function
 * @name deserialize
 * @description traverse js object and try convert values to their respective
 *              js type i.e numbers etc
 * @param  {Object} object valid js plain object
 * @return {Object}        object with all nodes converted to their respective
 *                                js types
 *
 * @since 0.1.0
 * @private
 */
exports.deserialize = function (object) {

  //ensure object
  object = _.merge({}, object);

  //TODO allow custom deserializer

  //traverse object and apply parser
  traverse(object).forEach(function (value) {

    //parse number strings to js numbers
    const isNumber = !isNaN(value);
    if (isNumber) {
      value = Number(value);
      //update current field
      this.update(value);
    }

  });

  return object;
};


/**
 * @object
 * @name reds
 * @description redis search utility
 * @since 0.1.0
 * @public
 */
exports.reds = reds;


/**
 * @object
 * @name indexes
 * @description map of all collection search indexes
 * @since 0.1.0
 * @private
 */
exports.indexes = {};


/**
 * @function
 * @name client
 * @description obtain current active redis client
 * @return {Object} redis client
 * @since 0.1.0
 * @private
 */
exports.client = function () {
  //obtain current redis client
  const _client = redis.client();

  //set reds client
  if (!reds.client) {
    reds.client = _client;
  }

  return _client;
};


/**
 * @function
 * @name key
 * @description prepare storage key
 * @return {Object} redis client
 * @since 0.1.0
 * @private
 */
exports.key = function (...args) {
  const _key = redis.key([].concat(...args));
  return _key;
};


/**
 * @function
 * @name multi
 * @description create new instance of redis multi object
 * @return {Object} redis client
 * @since 0.1.0
 * @private
 */
exports.multi = function () {
  const client = exports.client();
  const multi = client.multi();
  return multi;
};


/**
 * @function
 * @name indexKey
 * @description prepare collection index key
 * @param  {String} collection valid name of the collection
 * @since 0.1.0
 * @public
 */
exports.indexKey = function (collection) {
  collection = [].concat(collection).concat(['search']);
  const indexKey = exports.key(collection);
  return indexKey;
};


/**
 * @function
 * @name save
 * @description store object into redis as hash
 * @param  {Object|Object[]}   objects objects to be stored
 * @param  {Object}   [options] options to be used on storage
 * @param  {Boolean}   [options.index] if index is allowed or not. default true
 * @param  {String[]}   [options.ignore] field to be excluded from search indexes
 * @param  {String}   [options.collection] collection name used in key 
 *                                         namespacing. default hash
 * @param  {Function} done   a callback to invoke on success or failure
 * @return {Object}          created and persisted object with key added
 * @since 0.1.0
 * @version 0.2.0
 * @public
 */
exports.save = exports.create = function (objects, options, done) {
  //TODO create and index operations atomically
  //TODO fork reds and allow to pass multi on its operations
  //TODO refactor

  //normalize arguments
  if (_.isFunction(options)) {
    done = options;
    options = {};
  }

  //merge with default options
  options = _.merge({}, {
    index: true,
    collection: 'hash',
    ignore: ignore
  }, options);

  //ensure _id is ignored on indexing
  options.ignore = _.uniq(ignore.concat(options.ignore));

  //collect for multi create
  objects = [].concat(objects);

  //ensure they are objects
  objects = _.map(objects, function (object) { return _.merge({}, object); });

  //start multi save
  const _client = exports.multi();

  //iterate & map
  objects = _.map(objects, function (object) {

    //ensure js types
    object = exports.deserialize(object);

    //TODO validate key to ensure it start with required prefix

    //ensure timestamps
    if (object._id) {
      //update timestamps
      object.createdAt = object.createdAt || Date.now();
      object.updatedAt = Date.now();
    } else {
      //set timestamps
      object.createdAt = object.updatedAt = Date.now();
    }

    //obtain key from the object to be saved
    //or generate one
    object._id = object._id || exports.key([options.collection, uuid.v1()]);

    //flat the object
    let flatObject = flat(object);

    //serialize flattened object
    flatObject = exports.serialize(flatObject);

    //ensure index
    if (options.index) {
      //prepare collection search index
      const indexKey = exports.indexKey(options.collection);
      exports.indexes[indexKey] =
        exports.indexes[indexKey] || exports.reds.createSearch(indexKey);


      //index flat object
      _.forEach(flatObject, function (value, key) {
        //ignore object key and ignored fields from indexes
        key = _.last(key.split('.')); //obtain actual field name after flatting the object
        if (!_.has(options.ignore, key)) {
          //ensure indexes
          value = String(value);
          exports.indexes[indexKey].index(value, flatObject._id);
        }
      });
    }

    //queue save
    _client.hmset(flatObject._id, flatObject);

    //collect mapped object
    return object;

  });

  //ensure single or multi objects
  objects = objects.length === 1 ? _.first(objects) : objects;

  //save the objects and flush indexes
  _client.exec(function afterSave(error) {
    done(error, objects);
  });

};


/**
 * @function
 * @name get
 * @description get objects from redis
 * @param  {String|String[]|...String}   keys a single or collection of existing keys
 * @param  {Object} [criteria] optional selection criteria
 * @param  {Array} [criteria.fields] optional fields to select
 * @param  {Function} done   a callback to invoke on success or failure
 * @return {Object|[Object]} single or collection of existing hash
 * @since 0.1.0
 * @public
 */
exports.get = function (...keys) {

  //TODO refactor to query?

  //normalize keys to array
  keys = [].concat(...keys);

  //TODO ensure they are valid keys

  //compact and ensure unique keys
  keys = _.uniq(_.compact(keys));

  //obtain callback
  const done = _.last(keys);
  //drop callback if provided
  if (_.isFunction(done)) {
    keys = _.initial(keys);
  }

  //obtain selection criteria if specified
  let criteria = _.last(keys);
  criteria = _.isPlainObject(criteria) ? criteria : undefined;
  //drop criteria if provided
  if (criteria) {
    //shape criteria to ensure fields are in array
    criteria.fields = query.fields(criteria.fields);
    keys = _.initial(keys);
  }
  //check if there are fields
  const criteriaHasFields = (criteria && criteria.fields);

  //initiate multi command client
  const _client = exports.multi();

  //prepare multiple hgetall
  _.forEach(keys, function (key) {

    //select specified hash fields
    if (criteriaHasFields) { _client.hmget(key, criteria.fields); }

    //select all hash fields
    else { _client.hgetall(key); }

  });

  //execute batch / multi commands
  _client.exec(function (error, objects) {

    //unflat objects
    if (!error) {

      //unflatten objects
      objects = _.map(objects, function (object) {

        //unflatten object from redis
        object = unflat(object);

        //parse object
        object = exports.deserialize(object);

        //ensure selected fields name
        //NOTE: redis return fields in order as they have requested
        if (criteriaHasFields) {
          _.forEach(object, function (value, key) {
            const fieldName = criteria.fields[key];
            if (fieldName) {
              object[fieldName] = value;
              delete object[key];
            }
          });
        }

        //double unflat the object in case of 
        //specified fields for selections
        return unflat(object);

      });

      //ensure single or multi objects
      objects = keys.length === 1 ? _.first(objects) : objects;

    }

    //ensure objects exists
    if (_.isEmpty(objects)) {
      error = error || new Error();
      error.message = 'Not Found';
      error.status = 404;
      objects = undefined;
    }

    done(error, objects);

  });

};


/**
 * @function
 * @name remove
 * @description remove objects from redis
 * @param  {String|String[]|...String} keys  a single or collection of existing keys
 * @param  {Function} done   a callback to invoke on success or failure
 * @return {Object|[Object]} single or collection of existing hash
 * @since 0.3.0
 * @public
 */
exports.remove = function (...keys) {
  //TODO refactor to query?

  //normalize keys to array
  keys = [].concat(...keys);

  //compact and ensure unique keys
  keys = _.uniq(_.compact(keys));

  //obtain callback
  const done = _.last(keys);
  //drop callback if provided
  if (_.isFunction(done)) {
    keys = _.initial(keys);
  }

  //ensure keys
  if (keys && keys.length > 0) {
    //initiate multi command client
    const _client = exports.multi();

    //delete each keys in transaction
    _client.del(keys);

    //execute deletes in batch
    //TODO shape the returned response
    _client.exec(done);

  }

  //reply with bad request
  //as no keys specified
  else {
    let error = new Error('Missing Keys');
    error.status = 400;
    done(error);
  }

};


/**
 * @function
 * @name search
 * @description search stored objects from redis
 * @param  {Object|String}   options    search options
 * @param  {String}   options.collection    searched collections. default to hash
 * @param  {String}   options.type    search operator(and / or). default to or
 * @param  {String}   options.q    search term. default to ''
 * @param  {String|String[]}   options.fields    fields to select
 * @param  {Function} done   a callback to invoke on success or failure
 * @since 0.1.0
 * @public
 */
exports.search = function (options, done) {
  //normalize options
  if (_.isString(options)) {
    options = {
      q: options
    };
  }

  //merge options
  options = _.merge({}, {
    type: 'or', //default search operator
    collection: 'hash', //default collection or namespace
    q: '' //search term
  }, options);

  //obtain search index
  const indexKey = exports.indexKey(options.collection);
  const search = exports.indexes[indexKey];

  //perform search if search index exists
  //TODO try using multi or lua script
  if (search && !_.isEmpty(options.q)) {
    async.waterfall([
      function find(next) {
        //issue search query using reds
        search.query(options.q).type(options.type).end(next);
      },
      function onSearchEnd(keys, next) {
        if (options.fields) {
          exports.get(keys, { fields: options.fields }, next);
        } else {
          exports.get(keys, next);
        }
      },
      function normalizeResults(results, next) {
        results = [].concat(results);
        next(null, results);
      }
    ], function onFinish(error, results) {
      done(error, results);
    });
  }

  //no search index exists return empty results
  else {
    done(null, []);
  }
};


//TODO pagination
//TODO count
//TODO sorting
//TODO add ability to score base on fields
//TODO add timestamps(createdAt, updatedAt)
//TODO implement remove
//TODO metadata
//TODO support unique secondary indexes
//TODO ensure id's are unique(i.e not exists)
//TODO build metadata if not provided
//TODO benchmarks
//TODO improve test coverage
//TODO notify after create, update, delete